File: src/shared/ipc-message.ts
import { v4 as uuidv4 } from 'uuid'

export enum IPCMessageType {
    AGENT_RESPONSE,
    AGENT_QUERY
}

export class IPCMessage<MessageBodyType = any> {
    type: IPCMessageType
    body?: MessageBodyType
    id: string

    constructor(type: IPCMessageType, body?: any, id: string = uuidv4().replaceAll(/[^0-9a-fA-F]/g, "")) {
        this.type = type
        this.body = body
        this.id = id
    }

    serialize() {
        const id = Buffer.from(this.id, "hex")

        const descriptor = new Uint8Array(2)
        descriptor[0] = 0b00000000 | this.type  // Highest bit indicates whether more packets exist (unused), lower bits indicate message type
        descriptor[1] = id.length // Indicates id buffer length

        const data = [descriptor, id]
        if(this.body) {
            data.push(Buffer.from(JSON.stringify(this.body), "utf-8"))
        }

        return Buffer.concat(data)
    }

    static deserialize<MessageBodyType>(data: Buffer) {
        const descriptor = data.slice(0, 2)
        const id = data.slice(2, 2+descriptor.at(1)!)
        const body = data.slice(2+descriptor.at(1)!)

        let bodyDecoded: string | undefined = body.toString("utf-8")
        if(bodyDecoded == "") bodyDecoded = undefined
        else bodyDecoded = JSON.parse(bodyDecoded)
        
        return new IPCMessage<MessageBodyType>(descriptor.at(0)! & 0b01111111, bodyDecoded, id.toString("hex"))
    }
}

File: src/agent/ipc/ipc_message.py
import json
import string
import uuid
import re

class IPCMessage():
    type: int
    body: str
    id: str

    def __init__(self, type: int, body: str = None, id: str = re.sub(r'[^A-Fa-f0-9]', "", str(uuid.uuid4()))):
        self.type = type
        self.body = body
        self.id = id

    def toJSON(self):
        return json.dumps(self, default=lambda o: o.__dict__)

    @staticmethod
    def deserialize(data: bytes):
        more_packets = bool(data[0] & (1 << 7))  # true if more packets exist for same message
        type = data[0] & 0b01111111
        id = data[2:2+data[1]].hex()
        if(len(data) <= 2+data[1]):
            body = None
        else:
            body = data[2+data[1]:].decode("utf-8")
        
        return IPCMessage(type, json.loads(body), id)

    def serialize(self):
        id_bytes = bytearray.fromhex(self.id)
        descriptor = [0b00000000 & self.type, len(id_bytes)]
        return bytearray(descriptor) + id_bytes + json.dumps(self.body).encode("utf-8")

